### ISSUE TITLE
Architecture - Essential - "at-least-once" TransferProcessListener

### ISSUE THREAD

**Issue by marcgs** [eclipse-dataspaceconnector/DataSpaceConnector#434]
Triggering custom logic on TransferProcess state changes requires registering a [TransferProcessListener](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/spi/src/main/java/org/eclipse/dataspaceconnector/spi/transfer/TransferProcessListener.java). However listeners are not guaranteed to be called, e.g. if the instance crashes after persisting the state change but before invoking the listener. Custom logic in the listener can be crucial for the functioning of client applications, thus EDC needs to guarantee that these are called before advancing a transfer process to the next state.
 
Suggestion for improvements:
* Add additional TransferProcess states to indicate that a TP has advanced state but listeners have not yet been called.
* Add processing for these new states in the TransferProcessManager, ensuring that listeners are always called.
* (Optional) Consider introducing a retry strategy for failing listeners
* (Optional) Consider executing listeners in parallel to improve performance

**Comment by jimmarino** [comment]
Can this issue be solved by calling the listeners before the state change? This could involve defining before and after change events?

I would be hesitant to add additional states for this as calling listeners seems to be an implementation detail and not a state transition itself. 

Couldn't the optional items be implemented using strategies in the listeners themselves so that this complexity is not introduced into the runtime? For example, listeners by design should execute quickly and if they take a while to complete, need to implement appropriate mechanisms such as a work queue.

**Comment by marcgs** [comment]
For our concrete use case it would make no difference to have the listeners called before state change (not sure if there are use case where it would though). The [current API](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/spi/src/main/java/org/eclipse/dataspaceconnector/spi/transfer/TransferProcessListener.java#L28) specifies that the callback is done after the state change is completed, this would need to be modified. 

We could certainly extend it with before/after change event callbacks too, but then we run into the same issue for the "after change" callbacks. I understand your concerns of overbloating transfer process states. Do you see an alternative solution for guaranteeing execution?

Regarding the optional items, I think we can certainly go this route if we document it this way. WDYT @algattik ?

**Comment by algattik** [comment]
Indeed calling the listeners after persisting the state change should resolve the issue. This would ensure that if the manager loop crashes before persisting the state change, and the process restarts, we are guaranteed that the listeners are called again (potentially more than once).

I don't think *after state change* events are needed in addition.

**Comment by ndr-brt** [comment]
I wouldn't change the order of the execution, `TransferProcessListener` has methods expressed in the past tense like `created` and `provisioned`, so they should be executed after the state change, otherwise this will introduce confusion.
In my opinion the cleanest approach is to introduce events, that could also become the source of truth for `TransferProcess` entity.

**Comment by algattik** [comment]
The methods could be renaming to `...ing`. What do you mean by introduce events, isn't that essentially the payload to these listeners? How would the implementation of at-least-once differ?

**Comment by ndr-brt** [comment]
yes, they could, but a listener should listen to something that happened, not something that is going to happen, that is confusing...
what does it mean it's `provisioning`? the provision is started? some resources are provisioned? something went wrong?
I would change that in `provisionTriggered` or `provisionStarted`, that is a clear statement: something happened.

Using events would help for the "at-least-once" because the event will be persisted and could be retrieved by the listener in any moment, also after a crash. Every event should be versioned, and the events on a single TransferProcess could also be considered as the "source of truth" for the entity itself (read: event sourcing).

**Comment by algattik** [comment]
So you mean the listener would be called on another thread, right?

**Comment by ndr-brt** [comment]
Yes, could be also by another instance of the connector.

**Comment by algattik** [comment]
We would still have to send the event to the event persistence store before persisting the transfer state. As I understand @jimmarino's comment, the idea is that the implementor could plug such a mechanism themselves. It seems there is even such an implementation for Azure Event Grid in the codebase already.

**Comment by DominikPinsel** [comment]
> Yes, could be also by another instance of the connector.

So in case this event is send to the EventGrid before the TransferProcess is stored, the other connector still could not tell with certainty whether the store wasn't updated in the meantime until he is able to process the Event?

**Comment by ndr-brt** [comment]
In my idea, an ordered list of events represent also the state change, so if the event is persisted, the state is changed, otherwise not.
The core could persist and publish those events to all the subscribers.

**Comment by ndr-brt** [comment]
> > Yes, could be also by another instance of the connector.
> 
> So in case this event is send to the EventGrid before the TransferProcess is stored, the other connector still could not tell with certainty whether the store wasn't updated in the meantime until he is able to process the Event?

The event would be published only after its persistence (that will determine the state change). The event grid will become overkill, because the `TransferProcessStore` will already contain all the events.
Those are the basics of the "Event sourcing" pattern.

**Comment by ndr-brt** [comment]
> > The event would be published only after its persistence (that will determine the state change).
> 
> How would that solve the reported issue? The state could could be persisted and then the application could crash before being able to publish the event.

The listeners would have the possibility to retrieve the events that are still not processed

**Comment by jimmarino** [comment]
If there has to be guaranteed synchronization between a listener and a state change, both operations will need to be performed within the context of a single "transaction". That would entail using a single transactional resource such as a JDBC connection or using XA transactions. Since very few storage and queuing systems properly support XA, that's probably not something we can rely on. 

I'm also concerned this will introduce a lot of complexity into the EDC.

Maybe this issue can be solved in a different way. @marcgs could you outline some of the specific cases you guys ran into that need to be solved? Perhaps there is an alternative. For example, if an action needs to be performed prior to dispatching a request from the connector, a `Provisioner` implementation could be written to handle that. This would be called reliably.

Another alternative would be to have a trigger or watchdog mechanism fire from the database.

Also, do we really need this level of reliability? For example, even if an event is persisted, there is no (easy) way to guarantee that those events will be handled in an ordered fashion or that the transfer process will be in the same state by the time the event is processed (e.g. another state transition could happen before the event is processed).

**Comment by algattik** [comment]
There is no need for transactional behavior, i.e. properties such as atomicity, isolation are not required.

What we do need is a lower level of guarantee: to ensure our callback code can be called **at-least-once** when e.g. a transfer completes. We need to run business logic on the Consumer side when the Provider has finished transferring data (in our case, to initiate new calls to other Providers to retrieve data recursively from multiple data space partitions).

In this scenario we don't need guarantees that the transfer will not have progressed to another state, and an asynchronous mechanism is fine.

The problem with the current behavior is that the framework can persist the fact that the transfer has completed and then never call our callback code.

Currently the `AsyncTransferProcessManager` code:
1. changes the state `process.transition...()`
2. persists the change `transferProcessStore.update(process)`
3. invokes listeners `invokeForEach(l -> l.created(process))`

If we just invert 2 and 3 I think we can achieve the desired outcome, without changing any operation semantics.

**Comment by jimmarino** [comment]
Based on what @algattik outlined, I think we should call the listeners before the state transition. This is also how most transactional listener mechanisms work (e.g. JTA's Synchronization interface which has before and after callbacks). Listener implementations can use their own reliability infrastructure if needed. Otherwise, we would need to define something equivalent to the XA protocol, which isn't practical. 

The semantics would be what Alex outlined, **at-least-once**, no ordering guarantees, and only the guarantee that an attempt will eventually be made to transition the processes to the next state (e.g. the transition may never succeed).

That sounds like it is sufficient for the vast majority of application-level uses cases. For those cases that require absolute guarantee, users can build their own reliability mechanism.

**Comment by algattik** [comment]
Calling the listeners before the state transition, as you propose, would work. However, it would also work to keep calling the listeners after the state transition, as long as they are called *before the transition is persisted *. That latter path would have the advantage of retaining the existing interface.

**Comment by ndr-brt** [comment]
But, in that case how could we handle situations where the listeners are called but for some reasons the state transition doesn't succeed?

**Comment by algattik** [comment]
@ndr-brt Persisting the transition change may fail, in that case the listener will be called again (hence the request for "at-least-once" behavior)
